File: algorithms/fibonacci/fibonacci.js
function fibonacciRecursive(n) {
  if (n <= 1) {
    return n;
  }
  return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}


function fibonacciIterative(n) {
  if (n <= 1) {
    return n;
  }
  
  let a = 0;
  let b = 1;
  
  for (let i = 2; i <= n; i++) {
    const temp = a + b;
    a = b;
    b = temp;
  }
  
  return b;
}


function fibonacciMemoized(n, memo = {}) {
  if (n in memo) {
    return memo[n];
  }
  
  if (n <= 1) {
    return n;
  }
  
  memo[n] = fibonacciMemoized(n - 1, memo) + fibonacciMemoized(n - 2, memo);
  return memo[n];
}

// Export functions for testing
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    fibonacciRecursive,
    fibonacciIterative,
    fibonacciMemoized
  };
}

File: algorithms/fibonacci/fibonacci.rb
def fibonacci_recursive(n)
  return n if n <= 1
  fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)
end


def fibonacci_iterative(n)
  return n if n <= 1
  
  a, b = 0, 1
  (2..n).each do
    a, b = b, a + b
  end
  
  b
end


def fibonacci_memoized(n, memo = {})
  return memo[n] if memo.key?(n)
  return n if n <= 1
  
  memo[n] = fibonacci_memoized(n - 1, memo) + fibonacci_memoized(n - 2, memo)
end
